### ISSUE TITLE
Transport Layer Security

### ISSUE THREAD

**Issue by xj9** [Bitmessage/PyBitmessage#264]
Currently peer-to-peer communication occurs over insecure sockets. Wrapping these sockets in SSL is done via `ssl.wrap_socket`, the resulting object has most of the same methods as a standard socket.

Each peer could generate and self-sign a certificate and be configured to accept any certificate that it is presented. (We don't really care who we're connecting to as long as its over a secure channel, for now)

**Comment by xloem** [comment]
I don't think this old thread is quite right.

BM *does* authenticate the sender of a message.  Every message is signed by senders signing key.  Only the sender has the signing key.  Only the recipient can read the message.  The recipient can send it to someone else and sign it themselves, but that is more "forwarding mail" and less "replay attack".

If somebody can (A) sniff both your incoming and outgoing traffic (an ISP or a Sybil) and (B) read a message you send (be the recipient or compromise the device of a recipient), *then* they can identify that your IP sent that message from `outgoing - incoming`.  But SSL won't help much with this!  You'll still push more traffic when you send, and Sybil decrypts your traffic to forward it.  Tor will help more but still won't solve the problem, the tor project itself says it is weak in the face of that kind of power.

But couldn't you change the default client behavior (not even the protocol) to fix this?  Send out new objects at regular intervals and then `outgoing - incoming = constant`.  When a legitimate message is ready, send it at the next interval.  Bam.  No traffic change, even at the protocol level.

**Comment by Kagami** [comment]
@xloem see [comment by @marcelklehr](https://github.com/Bitmessage/PyBitmessage/issues/264#issuecomment-45957740). TLS will help against passive adversary. But won't help against MITM obviously.

Constant outgoing traffic rate (I guess this is similar to cover traffic in projects like I2P) won't help because protocol is open and you still can perfectly see which objects were incoming and which are created by the node.

Bitmessage whitepaper actually contains one solution against global adversary (see page 4) - by putting objects into acknowledgement field, but it hasn't been implemented in clients yet.

**Comment by PeterSurda** [comment]
I have briefly looked at how TLS, OpenSSL and how it works with python. I think we can use a certificate-less (anonymous) cipher, for example: AECDH-AES256-SHA (since Bitmessage already uses ECDH). OpenSSL seems to support this. This means there is no need to fiddle around with the certificates. The nodes do not authenticate, which actually protects them from being identified without having to generate certificates for each connection (which consumes unnecessary resources). It can work exactly the same way whether you use TOR or not. Let me know what you think about it.

**Comment by PeterSurda** [comment]
I spent some time on this. I have had quite a lot of success. I haven't committed the code because it needs to be cleaned up and further tested, but it looks like it's working.

I implemented a new bit in the services flags, 2, to indicate SSL support (1 is already taken but it doesn't appear to do anything). After exchanging versions, if a node sees that the other node supports encryption, it does a TLSv1 handshake with AECDH-AES256-SHA as I envisioned, and afterwards the sender/receiver threads switch to communicating through the SSL socket instead of the normal TCP socket. Unfortunately the handshake needs some workarounds due to deficiencies in how python implements ssl.wrap_socket. Mainly, the server needs to be provided a dummy key and a certificate which it loads even though it does not actually use them.

However, after the SSL handshake is done, they two nodes communicate via SSL. I analysed it through wireshark, and indeed they do exchange ECDSA public keys and then the rest goes through the tunnel.

If the connected node does not have the service flag 2, it behaves as before.

Let me know what you think about it.
